sound/soc/sof/loader.c

```diff
@@ -873,6 +873,15 @@ int snd_sof_run_firmware(struct snd_sof_dev *sdev)
 		return ret;
 	}
 
+	mutex_lock(&sdev->cores_status_mutex);
+	/*
+	 * fw booted, Only the master DSP core (0) should be ON here,
+	 * update to reflect it.
+	 */
+	sdev->core_refs[0] = 1;
+	sdev->enabled_cores_mask = BIT(0);
```

Comment:
I wonder if you need to remove the mask to replace it with a refcount. Having two separate fields seems like a recipe for race conditions and misaligned setups.

Reply:
> I wonder if you need to remove the mask to replace it with a refcount. Having two separate fields seems like a recipe for race conditions and misaligned setups.

That could be done, e.g. via helper like:
```
int get_enabled_cores_mask(sdev)
{
    int enabled_cores_mask = 0;
    for (i = 0; i < MAX_CORE_NUM; i++)
        if (sdev->core_refs[i] > 0)
            enabled_cores_mask |= BIT[i];

    return enabled_cores_mask;
}
```

But since we update the core_refs[] and enabled_cores_mask only in get/put() with mutex hold, we can still make sure they are aligned. Keeping the enabled_cores_mask can provide quick query and benefits in case of the querying happens frequently.